File: lib/xairo/image_surface.ex
defmodule Xairo.ImageSurface do
  @moduledoc false

  defstruct [:surface]

  def new(format, width, height) do
    with {:ok, image_surface} <- Xairo.Native.image_surface_create(format, width, height) do
      %__MODULE__{
        surface: image_surface
      }
    end
  end

  def format(%__MODULE__{surface: surface}) do
    Xairo.Native.image_surface_format(surface)
  end

  def width(%__MODULE__{surface: surface}) do
    Xairo.Native.image_surface_width(surface)
  end

  def height(%__MODULE__{surface: surface}) do
    Xairo.Native.image_surface_height(surface)
  end

  def stride(%__MODULE__{surface: surface}) do
    Xairo.Native.image_surface_stride(surface)
  end

  def write_to_png(%__MODULE__{surface: surface}, filename) do
    Xairo.Native.image_surface_write_to_png(surface, filename)
  end
end

File: native/xairo/src/image_surface.rs
use crate::enums::Format;
use crate::error::XairoError;
use rustler::ResourceArc;
use std::fs::File;

pub struct ImageSurfaceRaw {
    pub surface: cairo::ImageSurface,
}

unsafe impl Send for ImageSurfaceRaw {}
unsafe impl Sync for ImageSurfaceRaw {}

pub type ImageSurface = ResourceArc<ImageSurfaceRaw>;

#[rustler::nif]
fn image_surface_create(
    format: Format,
    width: i32,
    height: i32,
) -> Result<ImageSurface, XairoError> {
    let format: cairo::Format = format.into();
    match cairo::ImageSurface::create(format, width, height) {
        Ok(surface) => Ok(ResourceArc::new(ImageSurfaceRaw { surface })),
        Err(err) => Err(err.into()),
    }
}

#[rustler::nif]
fn image_surface_format(surface: ImageSurface) -> Format {
    surface.surface.format().into()
}

#[rustler::nif]
fn image_surface_width(surface: ImageSurface) -> i32 {
    surface.surface.width()
}

#[rustler::nif]
fn image_surface_height(surface: ImageSurface) -> i32 {
    surface.surface.height()
}

#[rustler::nif]
fn image_surface_stride(surface: ImageSurface) -> i32 {
    surface.surface.stride()
}

#[rustler::nif]
fn image_surface_write_to_png(surface: ImageSurface, filename: String) -> Result<(), XairoError> {
    match File::create(filename) {
        Ok(mut file) => match surface.surface.write_to_png(&mut file) {
            Ok(_) => Ok(()),
            Err(_) => Err(XairoError::WriteError),
        },
        Err(_) => Err(XairoError::WriteError),
    }
}
